sdk/tests/js/tests/tex-image-and-sub-image-2d-with-image-data.js

```diff
@@ -107,30 +115,63 @@ function generateTest(internalFormat, pixelFormat, pixelType, prologue, resource
                        gl.TEXTURE_CUBE_MAP_POSITIVE_Z,
                        gl.TEXTURE_CUBE_MAP_NEGATIVE_Z];
         }
+        // Handle the source sub-rectangle if specified (WebGL 2.0 only)
+        if (sourceSubRectangle) {
+            gl.pixelStorei(gl.UNPACK_SKIP_PIXELS, sourceSubRectangle[0]);
+            gl.pixelStorei(gl.UNPACK_SKIP_ROWS, sourceSubRectangle[1]);
+        }
         // Upload the image into the texture
         for (var tt = 0; tt < targets.length; ++tt) {
-            if (useTexSubImage2D) {
-                // Initialize the texture to black first
-                gl.texImage2D(targets[tt], 0, gl[internalFormat], imageData.width, imageData.height, 0,
-                              gl[pixelFormat], gl[pixelType], null);
-                gl.texSubImage2D(targets[tt], 0, 0, 0, gl[pixelFormat], gl[pixelType], imageData);
+            if (sourceSubRectangle) {
```

Comment:
Looking at this again, could we define:
  var uploadWidth = imageData.width;
  var uploadHeight = imageData.height;
and override them if sourceSubRectangle is defined?

That way we could reference those variables in the tex{Sub}Image2D calls and avoid duplicated code.